src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -459,12 +461,30 @@ private String formatError(DynamicObject string) {
             return StringUtils.format("invalid byte sequence in %s", Layouts.STRING.getRope(string).getEncoding());
         }
 
+        @Specialization(guards = {
+                "isInitialized(regexp)",
+                "isRubyString(string)",
+                "!isRubyEncoding(isCompatibleNode.executeCompatibleQuery(regexp, string))" })
+        protected Object searchRegionIncompatibleEncoding(DynamicObject regexp, DynamicObject string, int start, int end,
+                boolean forward,
+                @Cached ToEncodingNode toEncodingNode,
+                @Cached("toEncodingNode.executeToEncoding(regexp)") Encoding regexpEncoding,
+                @Cached("toEncodingNode.executeToEncoding(string)") Encoding stringEncoding,
+                @Cached EncodingNodes.CompatibleQueryNode isCompatibleNode) {
+            throw new RaiseException(getContext(), coreExceptions().encodingCompatibilityErrorIncompatible(regexpEncoding, stringEncoding, this));
+        }
+
         @Specialization(
-                guards = { "isInitialized(regexp)", "isRubyString(string)", "isValidEncoding(string, rangeNode)" })
+                guards = {
+                        "isInitialized(regexp)",
+                        "isRubyString(string)",
+                        "isValidEncoding(string, rangeNode)",
+                        "!isNil(isCompatibleNode.executeCompatibleQuery(regexp, string))" })
```

Comment:
In general I find it useful to have the same guard expression in both specializations, one negated and one not.
Could you use:
```
isRubyEncoding(isCompatibleNode.executeCompatibleQuery(regexp, string))
```
here?

Reply:
That is what I had before 8261aba3028e16b1bb31348351451ffcbd94bbb3. I thought it would be cheaper to check if the return is nil than checking if it is an encoding. I can remove that commit if you think we should use the same guard expression.

Reply:
I'd go for the I'd go for @eregon's suggestion. Specializations get tricky to reason about very quickly and keeping them regular really helps.